### PR TITLE
Inconsistent behavior for parsing roles/members from ldap memberOf

### PR DESCRIPTION
**Pull request by andythsu**
## Description

This should resolve https://github.com/trinodb/trino-gateway/issues/308 if I'm not mistaken. Otherwise can close this PR.

## Additional context and related issues

Please see the linked issue for detailed explanation of the issue.

## Release notes

(X) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
* 
```

### CODE REVIEW

**Comment by andythsu** [comment]
@cghyzel didn't know you had a PR open for this. Let me know what you think

**Comment by Chaho12** [comment]
@cghyzel would this resolve your issue too?

**Review by Chaho12 — COMMENTED** [review]
have you seen the logs? even if i am admin in web, log keep saying that.
```
2024-04-09T11:22:28.196+0900    INFO    pool-6-thread-6 - POST /userinfo        io.trino.gateway.ha.security.LbAuthorizer       User 'hive' with memberOf(Optional[uid:hive]) identified as USER(^.+$)
2024-04-09T11:22:29.891+0900    INFO    pool-6-thread-11 - POST /webapp/getDistribution io.trino.gateway.ha.security.LbAuthorizer       User 'hive' with memberOf(Optional[uid:hive]) identified as USER(^.+$)

```

```
authorization:
  admin: "hive"
  user: "^.+$"
  api: "hive"
  ldapConfigPath: ldapConfig.yaml
```

**Comment by andythsu** [comment]
This doesn't mean your privilege. It means the operation you are performing belongs to USER operation. See https://github.com/trinodb/trino-gateway/blob/ac407dcbd02c3e00655a2ce0a63dd21be429585b/gateway-ha/src/main/java/io/trino/gateway/ha/security/LbAuthorizer.java#L38 this is doing a condition on `role`. This `role` is the privilege you need to perform the intended operation

**Comment by Chaho12** [comment]
> This doesn't mean your privilege. It means the operation you are performing belongs to USER operation

Thx for explanation :) 
```
2024-04-09T12:30:59.512+0900    INFO    pool-6-thread-8 - POST /webapp/saveBackend      io.trino.gateway.ha.security.LbAuthorizer       User 'hive' with memberOf(Optional[uid: hive]) was identified as ADMIN((.*)hive(.*))
```

**Comment by oneonestar** [comment]
I think #310 is a correct but incomplete fix.
#310 only fixed `/userinfo`, but `/findQueryHistory` also needs a fix.

```
        // findQueryHistory before fix
        String[] roles = principal.getMemberOf().orElse("").split("_");
        if (!Arrays.asList(roles).contains("ADMIN")) {
            query.setUser(principal.getName());
        }

        // can be fixed by changing it to:
        if (securityContext.isUserInRole("ADMIN")) {
            roles.add("ADMIN");
        }  
```

After both fixes landed, only LbAuthorizer will call principal.getMemberOf() and perform a mapping between memberOf and role.

File: gateway-ha/src/main/java/io/trino/gateway/ha/security/LbLdapParser.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        this.authorizationConfiguration = authorizationConfiguration;
+    }
+
+    /**
+     * The input is the ldap search from the user. It's in the format of
+     * String memberOf = """
+     * memberOf: ...
+     * memberOf: ...
+     * """
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/security/LbLdapParser.java:35]: Am not a ldap expert, but how are you sure that it is `memberOf: {...}` format?
I think this depends on how LDAP server returns with how we set ldapGroupMemberAttribute. 


`ldapGroupMemberAttribute: 'ou'` shows `ou: memberof={}`
image

`ldapGroupMemberAttribute: 'uid'` shows `uid: {}`
image

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/security/LbLdapParser.java:35]: right... there's some assumption made in this

**Pull request closed without merging**